subsys/net/lib/sockets/sockets.c

```diff
@@ -1344,8 +1344,13 @@ static inline ssize_t zsock_recv_dgram(struct net_context *ctx,
 			 * if that was set using connect() call.
 			 */
 			if (ctx->flags & NET_CONTEXT_REMOTE_ADDR_SET) {
+#if defined(CONFIG_NET_OFFLOAD)
+				memcpy(src_addr, &pkt->remote,
```

Comment:
I don't think `NET_CONTEXT_REMOTE_ADDR_SET` check makes any more sense in case address is provided via packet. And as a matter of fact, I think we should keep both methods for getting address to keep offloaded devices that do not set the address on packet still functional (or, well, semi-functional as this old menthod is not perfect).

What I'd suggest to do:
* Move the address setting on an offloaded interface to a function, as it's done for native stack (`sock_get_offloaded_pkt_src_addr()` for example)
* In case offloading is disabled, provide dummy (empty) implementation so that we can still use `IS_ENABLED()` above and build w/o errors.
* For the offloaded case:
  * Check if the `remote` was set on a packet (is not unspecified). If so, copy it,
  * If not, to fallback the old method.
  * If neither of those give results, return an error.

CC @jukkar In case you have any other opinion on this topic.

Reply:
I am a bit confused @jukkar and @rlubos . 

I have noticed that in `struct net_context` there is field `struct sockaddr remote`. Why cant the offloaded driver store the peer address there and set `NET_CONTEXT_REMOTE_ADDR_SET`? Then we don't need to modify `zsock_recv_dgram()` here or add a sockaddr in `struct net_pkt` right?

Or is this `struct sockaddr remote` in `struct net_context` not allowed to be used this way?

Reply:
I see at least 2 problems with such approach:
* `net_contect->remote` is used to store the peer address upon `net_context_connect()` call (and thus socket `connect()`). If we allow the offloaded driver to overwrite this information, we could no longer use `send()` reliably.
* And even if that was not an issue (`connect()`/`send()` not used), there's still a problem if you receive datagrams from  different peers. If you queue two `net_pkt`s at the socket level, how would you retrieve the address for both? Upon reception of the second packet, the remote address of the first one would be lost.

Reply:
I see, thank you for clearing it up for me. I'm almost done with your comment and will push soon.

Reply:
I have updated this with your comments.